WebXR should respond to changes in frame size

grorg opened the pull request:
#### 2da8807b15a6fe8b49b1b83484daa350c532001e
<pre>
WebXR should respond to changes in frame size
<a href="https://bugs.webkit.org/show_bug.cgi?id=246496">https://bugs.webkit.org/show_bug.cgi?id=246496</a>
&lt;rdar://problem/101151986&gt;

Reviewed by Antoine Quint.

The size of the framebuffer provided by the compositor may change during the
life of a session. Detect this and respond to it.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame): Look at the size of the incoming frame.
If it has changed, re-run the setupFramebuffer logic.
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer): Make sure we delete existing
WebGL objects if we create new ones.

Canonical link: <a href="https://commits.webkit.org/255543@main">https://commits.webkit.org/255543@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d5f0461af139f7f78bd4250fca4ee0fc0bebf17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23470 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102608 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96885 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2097 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30435 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85282 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98547 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79370 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28350 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36835 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34636 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3839 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38507 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40423 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/37348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -261,6 +275,12 @@ bool WebXROpaqueFramebuffer::setupFramebuffer()
     }
 
     if (m_attributes.antialias && m_context.isWebGL2()) {
+        if (m_resolvedFBO)
+            gl.deleteFramebuffer(m_resolvedFBO);
```
If only there was some kind of automatic GL object helper so we couldn't forget to do this.

djg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -325,9 +349,12 @@ bool WebXROpaqueFramebuffer::setupFramebuffer()
 
-    return true;
+    // FIXME: It seems unlikely we'd ever get to this point.
+    return false;
```
Should we just return `gl.checkFramebufferStatus(GL::FRAMEBUFFER_ == GL::FRAMEBUFFER_COMPLETE` instead?

grorg replied:
Yeah. I actually was tempted to put an ASSERT in there just to see if there ever was a situation we'd hit this.

graouts replied:
It seems like an `ASSERT_NOT_REACHED()` is in order.

grorg replied:
I'm not sure if other ports ever get into this state. I guess we'll find out because I just `return false;`ed them :). I'll add the `checkFramebufferStatus` though.

graouts commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -92,6 +92,25 @@ void WebXROpaqueFramebuffer::startFrame(const PlatformXR::Device::FrameData::Lay
+        return;
+
+    uint32_t bufferWidth = static_cast<uint32_t>(surfaceWidth);
+    uint32_t bufferHeight = static_cast<uint32_t>(surfaceHeight);
```
`auto`

grorg committed:
WebXR should respond to changes in frame size
https://bugs.webkit.org/show_bug.cgi?id=246496
<rdar://problem/101151986>

Reviewed by Antoine Quint.

The size of the framebuffer provided by the compositor may change during the
life of a session. Detect this and respond to it.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame): Look at the size of the incoming frame.
If it has changed, re-run the setupFramebuffer logic.
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer): Make sure we delete existing
WebGL objects if we create new ones.

Canonical link: https://commits.webkit.org/255543@main
